src/Database/LSMTree/Internal/RunReader.hs

```diff
@@ -96,6 +105,7 @@ new !offsetKey readerRun = do
 
     (page, entryNo) <- seekFirstEntry readerKOpsHandle
 
+    readerBlobFile <- dupRef runBlobFile
```

Comment:
Oh right because you could have a cursor and it might give you a blobref for the final entry in the table and then finalise the cursor, invalidating the blobref (unless the caller has a table handle open too).

Perhaps it'd be more uniform to have an explicit close for cursors that you _must always_ call, even if you fully drain the cursor. And then have the blob ref rule be that the cursor has to be still live, rather than requiring the table the cursor was derived from.

Reply:
Wasn't that already how it worked? I haven't looked at the diff of this PR yet whether anything in that regard has changed, but on `main` the cursor keeps a reference to all the runs, and those references are only removed on cursor close. That means blob references live precisely as long as the cursor.

Reply:
We have two consumers of `RunReaders`. They both hold all references to the runs being read until some defined "end":

1. `Cursor` already stays open until it is closed explicitly, at which point it releases the references to the runs (as you describe it should work), so any blob refs returned might get invalidated. (seems like that's currently not documented on `readCursor`, we should do that)
2. `MergingRun` stays open until the merge is complete, at which point it releases the references to the runs. We don't need them any more, since all blobs have already been written to the new run.

However, the `RunReader`s inside only stay around until their last entry has been read.

This is mainly due to the design of `MutableHeap`, which only contains non-empty readers (we can only (re-)insert if we have an entry at hand to `compare` on).  So it feels natural to close the readers the moment they become drained and can't be re-inserted. Otherwise we'd need to store them in some auxiliary structure to close them later.